src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherBuilder.cs

```diff
@@ -437,5 +499,57 @@ private static bool HasAdditionalRequiredSegments(MatcherBuilderEntry entry, int
 
             return false;
         }
+
+        private void ApplyPolicies(DfaNode node)
+        {
+            if (node.Matches.Count == 0)
+            {
+                return;
+            }
+
+            // Start with the current node as the root.
+            var work = new List<DfaNode>() { node, };
+            for (var i = 0; i < _nodeBuilders.Length; i++)
+            {
+                var nodeBuilder = _nodeBuilders[i];
+
+                // Build a list of each 
+                var nextWork = new List<DfaNode>();
+
+                for (var j = 0; j < work.Count; j++)
+                {
+                    var parent = work[j];
+                    if (!nodeBuilder.AppliesToNode(parent.Matches))
+                    {
+                        // This node-builder doesn't care about this node, so add it to the list
+                        // to be processed by the next node-builder.
+                        nextWork.Add(parent);
+                        continue;
+                    }
+
+                    // This node-builder does apply to this node, so we need to create new nodes for each edge,
+                    // and then attach them to the parent.
+                    var edges = nodeBuilder.GetEdges(parent.Matches);
+                    for (var k = 0; k < edges.Count; k++)
+                    {
+                        var edge = edges[k];
+
+                        var next = new DfaNode();
+                        next.Matches.AddRange(edge.Endpoints.Cast<MatcherEndpoint>().ToArray());
```

Comment:
These endpoints will always be `MatcherEndpoint`? No need to type check and provide a better error message?

Reply:
Related to https://github.com/aspnet/Routing/issues/648 - I'll leave a comment.